### PR TITLE
Exclude foldable expressions from GPU if constant folding is disabled

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds documentation that we don't support excluding the rule `org.apache.spark.sql.catalyst.optimizer.ConstantFolding`

This came up as a result of trying to support #1736 which was because of a change in Spark 3.1

Signed-off-by: Raza Jafri <rjafri@nvidia.com>

### CODE REVIEW

**Comment by razajafri** [comment]
build

File: docs/compatibility.md [hunk 1]
```diff
@@ -352,4 +352,10 @@ Casting from string to timestamp currently has the following limitations.
 - <a name="Footnote1"></a>[1] The timestamp portion must be complete in terms of hours, minutes, seconds, and
  milliseconds, with 2 digits each for hours, minutes, and seconds, and 6 digits for milliseconds. 
  Only timezone 'Z' (UTC) is supported. Casting unsupported formats will result in null values. 
- 
+
+### Constant Folding
+
+ConstantFolding is an operator optimization rule in Catalyst that replaces expressions that can
+be statically evaluated with their equivalent literal values. We don't support excluding the rule 
+`org.apache.spark.sql.catalyst.optimizer.ConstantFolding` as we rely on this rule internally
```
Comment (jlowe) [inline docs/compatibility.md:360]: Rather than just squirreling this away in the corner of a doc most users will not read, why not also update the plugin to check this config setting when it starts to optimize a plan?  It could emit a warning and refuse to optimize the plan which will let the query run correctly, albeit completely on the CPU.

File: docs/compatibility.md [hunk 2]
```diff
@@ -352,4 +352,10 @@ Casting from string to timestamp currently has the following limitations.
...
  milliseconds, with 2 digits each for hours, minutes, and seconds, and 6 digits for milliseconds. 
  Only timezone 'Z' (UTC) is supported. Casting unsupported formats will result in null values. 
- 
+
+### Constant Folding
+
+ConstantFolding is an operator optimization rule in Catalyst that replaces expressions that can
+be statically evaluated with their equivalent literal values. We don't support excluding the rule 
+`org.apache.spark.sql.catalyst.optimizer.ConstantFolding` as we rely on this rule internally 
+and excluding it will cause some of our assumptions to fail.
```
Comment (jlowe) [inline docs/compatibility.md:361]: I suggest rewording this to remove the assumptions part and simply say the query may execute incorrectly (or even crash if that's possible?)

File: docs/compatibility.md [hunk 3]
```diff
@@ -352,4 +352,10 @@ Casting from string to timestamp currently has the following limitations.
 - <a name="Footnote1"></a>[1] The timestamp portion must be complete in terms of hours, minutes, seconds, and
  milliseconds, with 2 digits each for hours, minutes, and seconds, and 6 digits for milliseconds. 
  Only timezone 'Z' (UTC) is supported. Casting unsupported formats will result in null values. 
- 
+
+### Constant Folding
+
+ConstantFolding is an operator optimization rule in Catalyst that replaces expressions that can
+be statically evaluated with their equivalent literal values. We don't support excluding the rule
```
Comment (jlowe) [inline docs/compatibility.md:359]: Suggest changing "We" to "The RAPIDS Accelerator" in this paragraph.

File: docs/compatibility.md (hunk 1 above)
Comment (revans2) [reply docs/compatibility.md:360]: The issue that would happen is that we get two literal values in say a GPU Add and we try to run it and it blows up. I agree that we probably want to disable the plugin, but at least output a warning if we see that ConstantFolding has been disabled.

As an alternative when replacing the plan with a GPU enabled version we could look for all foldable CPU operators, and if they are not on Literal values we CANNOT replace them.  Because then at least we could not crash, and we would honor what was asked by the end user, to disable constant folding.

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 4]
```diff
@@ -1028,6 +1028,10 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
     entry.get(conf)
   }
 
+  def getStringConf(key: String): Option[String] = {
+    conf.get(key)
+  }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1033]: I don't think we want to encourage code looking for confs in RapidsConf that aren't confs explicitly supported by it.  There are parts of the code that explicitly filters the Spark configs to `spark.rapids.*` before passing to code that will build a RapidsConf.

If we want to lookup a SQLConf key then we should use a SQLConf to do so.  `SQLConf.get` can be used to obtain the current active instance.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 5]
```diff
@@ -244,6 +245,18 @@ abstract class RapidsMeta[INPUT <: BASE, BASE, OUTPUT <: BASE](
       }
     }
 
+    val isExp = wrapped.isInstanceOf[Expression]
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:248]: I think this makes more sense to be in `BaseExprMeta` where we know we have an expression.  IIUC all expressions should be using it.

File: docs/compatibility.md [hunk 6]
```diff
@@ -352,4 +352,10 @@ Casting from string to timestamp currently has the following limitations.
...
  milliseconds, with 2 digits each for hours, minutes, and seconds, and 6 digits for milliseconds. 
  Only timezone 'Z' (UTC) is supported. Casting unsupported formats will result in null values. 
- 
+
+### Constant Folding
+
+ConstantFolding is an operator optimization rule in Catalyst that replaces expressions that can
+be statically evaluated with their equivalent literal values. The RAPIDS Accelerator doesn't
+support excluding the rule `org.apache.spark.sql.catalyst.optimizer.ConstantFolding` as we rely on 
+this rule internally and excluding it may cause the query to produce undesired results
```
Comment (jlowe) [inline docs/compatibility.md:361]: This is no longer true after the other changes in this file.  Instead this should mention parts of the query may not run on the GPU without constant folding optimization enabled.

**Review by jlowe — COMMENTED** [review]
The PR headline should be updated to reflect this is no longer just about updating docs.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala (hunk 5 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:248]: Also we have to allow Literal values to pass through this OK.  Literal values are foldable, but we support them

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:248]: isn't that covered by `exp.foldable && !GpuOverrides.isLit(exp)`. So the condition is true only if its foldable and not a literal.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 7]
```diff
@@ -244,6 +245,7 @@ abstract class RapidsMeta[INPUT <: BASE, BASE, OUTPUT <: BASE](
       }
     }
 
+
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:248]: nit extra blank line

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 8]
```diff
@@ -807,6 +809,16 @@ abstract class BaseExprMeta[INPUT <: Expression](
   }
 
   final override def tagSelfForGpu(): Unit = {
+    val isConstantFoldingOff =
+      SQLConf.get.optimizerExcludedRules
+        .contains("org.apache.spark.sql.catalyst.optimizer.ConstantFolding")
+    if (isConstantFoldingOff) {
+      val exp = wrapped.asInstanceOf[Expression]
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:816]: is this cast needed?

Comment (razajafri) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala]: no, I removed and  tested against Spark 3.0.0

**Comment by razajafri** [comment]
@jlowe @revans2 PTAL

File: docs/compatibility.md [hunk 9]
```diff
@@ -352,4 +352,10 @@ Casting from string to timestamp currently has the following limitations.
...
  milliseconds, with 2 digits each for hours, minutes, and seconds, and 6 digits for milliseconds. 
  Only timezone 'Z' (UTC) is supported. Casting unsupported formats will result in null values. 
- 
+
+### Constant Folding
+
+ConstantFolding is an operator optimization rule in Catalyst that replaces expressions that can
+be statically evaluated with their equivalent literal values. The RAPIDS Accelerator relies
+on constant folding and parts of the query will not be accelerated if 
+`org.apache.spark.sql.catalyst.optimizer.ConstantFolding` is excluded as a rule
```
Comment (jlowe) [inline docs/compatibility.md:361]:
```suggestion
`org.apache.spark.sql.catalyst.optimizer.ConstantFolding` is excluded as a rule.
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 4 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: A `RapidsConf` can be built from a `SQLConf` but _this is optional_.  That means one cannot assume that a `RapidsConf` contains `SQLConf` settings, and that's why getting arbitrary keys from a `RapidsConf` should not be supported.

**Review by jlowe — APPROVED** [review]
Looks good to me, but would be good to hear from @revans2 as well.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 10]
```diff
@@ -807,6 +808,15 @@ abstract class BaseExprMeta[INPUT <: Expression](
   }
 
   final override def tagSelfForGpu(): Unit = {
+    val isConstantFoldingOff =
+      SQLConf.get.optimizerExcludedRules
+        .contains("org.apache.spark.sql.catalyst.optimizer.ConstantFolding")
+    if (isConstantFoldingOff) {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:814]: nit: This looks great, but why do we need to check the config at all? `wrapped.foldable` should be cheap to do and it not directly impacted by the `optimizerExcludedRules`.

**Review by revans2 — APPROVED** [review]

**Comment by razajafri** [comment]
build

**Review by jlowe — APPROVED** [review]
